DSharpPlus.Core/DiscordClient.cs

```diff
@@ -0,0 +1,124 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// this is a stub, intended to demonstrate how to set up the rest client rather than
+// act as a working client
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+using DSharpPlus.Core.Rest;
+using DSharpPlus.Core.Rest.Ratelimiting;
+
+using Microsoft.Extensions.DependencyInjection;
+
+using Polly;
+using Polly.Contrib.WaitAndRetry;
+
+namespace DSharpPlus.Core
+{
+
+    public class DiscordClient
+    {
+        // unfortunately, we need those two.
+        public IServiceCollection ServiceCollection { get; private set; }
+
+        public ServiceProvider Services { get; private set; }
+
+        // this is what we end up using after init has completed.
+        public RestClient RestClient { get; private set; }
+
+        // again, one second
+        private static readonly TimeSpan oneSecond = TimeSpan.FromSeconds(1);
+
+        public DiscordClient(IServiceCollection serviceCollection)
+        {
+            ServiceCollection = serviceCollection;
+
+            // we need to register the rest client via DI to make sure everything works.
+            ServiceCollection
+                .AddSingleton<RestClient>()
+                .AddMemoryCache();
+
+            Services = serviceCollection.BuildServiceProvider();
+
+            #region set up policies, retrying etc. this should be moved to a separate method.
+            PollyRatelimitingPolicy ratelimiter = new();
+
+            // these values are arbitrary and should be configurable. they specify how much time should elapse between retries,
+            // and how often we should retry. the algorithm creates an exponentially increasing list of delays, with the first one
+            // being based on 0.5 seconds here.
+            IEnumerable<TimeSpan> retryDelay = Backoff.DecorrelatedJitterBackoffV2(TimeSpan.FromSeconds(0.5), 10);
+
+            ServiceCollection.AddHttpClient<RestClient>()
+                .ConfigureHttpClient((services, httpClient) =>
+                {
+                    httpClient.BaseAddress = new("https://discord.com/api/v9"); // or whatever, ideally have this in a constant
```

Comment:
All of this should be an extension method on the container, not in the constructor of a client.

Reply:
this is to show the concept, not to propose an implementation of the DiscordClient.